src/meta_memcache/base/memcache_socket.py

```diff
@@ -0,0 +1,278 @@
+                    # pyre-ignore[6]
+                    success.int_flags[f] = int(chunk[1:])
+                elif f := token_flags_values.get(flag):
+                    success.token_flags[f] = bytes(chunk[1:])
+                else:
+                    _log.warning(f"Unrecognized flag {bytes(chunk)}")
+
+    def _tokenize_header(self, header: memoryview) -> List[memoryview]:
+        """
+        Slice header by spaces into memoryview chunks
+        """
+        chunks = []
+        prev, i = 0, -1
+        for i, v in enumerate(header):
+            if v == SPACE:
+                if i > prev:
+                    chunks.append(header[prev:i])
+                prev = i + 1
+        if prev <= i:
+            chunks.append(header[prev:])
+        return chunks
+
+    def _get_header(self) -> List[memoryview]:
+        endl_pos = self._recv_header()
+        header = self._buf_view[self._pos : endl_pos]
+        self._pos = endl_pos + ENDL_LEN
+        return self._tokenize_header(header)
+
+    def sendall(self, data: Blob) -> None:
+        self._conn.sendall(data)
+
+    def get_response(
+        self,
+    ) -> Union[Value, Success, NotStored, Conflict, Miss]:
+        header = self._get_header()
+        response_code, *chunks = header
+        try:
+            if response_code == b"VA":
+                # Value response, parse size and flags
+                # pyre-ignore[6]
+                value_size = int(chunks.pop(0))
+                result = Value(value_size)
+                self._add_flags(result, chunks)
+            elif response_code == b"HD":
+                # Stored or no value, return Success
+                result = Success()
+                self._add_flags(result, chunks)
+            elif response_code == b"NS":
+                # Value response, parse size and flags
+                result = NotStored()
+                assert len(chunks) == 0  # noqa: S101
+            elif response_code == b"EX":
+                # Already exists, not changed, CAS conflict
+                result = Conflict()
+                assert len(chunks) == 0  # noqa: S101
+            elif response_code == b"EN" or response_code == b"NF":
+                # Not Found, Miss.
+                result = Miss()
+                assert len(chunks) == 0  # noqa: S101
+            else:
+                raise MemcacheError(f"Unknown response: {bytes(response_code)}")
+        except Exception as e:
+            raise MemcacheError(
+                f'Error parsing response header {b" ".join(header)}'
+            ) from e
+
+        self._reset_buffer()
+        return result
+
+    def get_value(self, size: int) -> Union[bytearray, memoryview]:
+        """
+        Get data value from the buffer and/or socket if hasn't been
+        read fully.
+        """
+        data_in_buf = self._read - self._pos
+        missing_data_size = size - data_in_buf
+        if missing_data_size > 0 and self._read < len(self._buf):
```

Comment:
One question, if we know the value size and that size is greater than our preassigned buffer, why don't just go directly to create a new one that holds the data instead of doing this first try?

Reply:
We discussed this in person, but for the record:
The header size is not known ahead of time. So when we read on the socket, we ask the operating system to fill the buffer we pass with all the data it has available. This might be just the header, the header + the full value, or the header and a chunk of the value (but not all).

Majority of small value requests will have header + whole value. Read in a single `recv` and written in a reusable buffer. Very efficient. If the data is large and doesn't fit in buffer, then we allocate a buffer for the value, copy the chunk in the read buffer (if any) and read the rest from the socket in blocking mode.